components/director/internal/model/formation_assignment.go

```diff
@@ -118,6 +132,69 @@ func (f *FormationAssignment) Clone() *FormationAssignment {
 	}
 }
 
+// IsInErrorState returns if the formation assignment is in error state
+func (f *FormationAssignment) IsInErrorState() bool {
+	state := f.State
+	return state == string(CreateErrorAssignmentState) ||
+		strings.HasSuffix(state, string(DeleteErrorAssignmentState))
+}
+
+// IsInProgressState returns if the formation assignment is in progress state
+func (f *FormationAssignment) IsInProgressState() bool {
+	return f.isInProgressAssignState() || f.isInProgressUnassignState()
+}
+
+// IsInRegularUnassignState returns if the formation assignment is in regular unassign stage
+func (f *FormationAssignment) IsInRegularUnassignState() bool {
+	unassignOperationStates := []string{string(DeletingAssignmentState),
+		string(DeleteErrorAssignmentState)}
+	return str.ValueIn(f.State, unassignOperationStates)
+}
+
+// SetStateToDeleting sets the state to deleting and returns if the formation assignment is updated
+func (f *FormationAssignment) SetStateToDeleting() bool {
+	if f.isInProgressUnassignState() {
+		return false
+	}
+	if strings.HasSuffix(f.State, string(DeleteErrorAssignmentState)) {
+		f.State = strings.Replace(f.State, string(DeleteErrorAssignmentState), string(DeletingAssignmentState), 1)
+		return true
+	}
+	f.State = string(DeletingAssignmentState)
+	return true
+}
+
+// GetOperation returns the formation operation that is determined based on the state of the assignment
+func (f *FormationAssignment) GetOperation() FormationOperation {
+	operation := AssignFormation
+	if f.State == string(DeleteErrorAssignmentState) || f.State == string(DeletingAssignmentState) ||
+		f.State == string(InstanceCreatorDeleteErrorAssignmentState) || f.State == string(InstanceCreatorDeletingAssignmentState) {
+		operation = UnassignFormation
+	}
+	return operation
+}
+
+// GetNotificationState returns the assignment state that is to be sent for notifications
+// and is exposed to the GraphQL layer
+func (f *FormationAssignment) GetNotificationState() string {
+	state := f.State
+	if strings.HasSuffix(state, string(DeleteErrorFormationState)) {
+		state = string(DeleteErrorFormationState)
+	} else if f.isInProgressUnassignState() {
+		state = string(DeletingAssignmentState)
+	}
+	return state
+}
+
+func (f *FormationAssignment) isInProgressAssignState() bool {
+	return f.State == string(InitialAssignmentState) ||
+		f.State == string(ConfigPendingAssignmentState)
+}
+
+func (f *FormationAssignment) isInProgressUnassignState() bool {
+	return strings.HasSuffix(f.State, string(DeletingAssignmentState))
```

Comment:
Will that catch the regular non-opt-in fearure deleting state?

Reply:
Yes, I tested it and `HasSuffix` matches `DELETING` state (`DELETING` is a suffix of `DELETING`). Also it is validated in the e2e tests via the FormationStatus asserts

Reply:
If that's the case, can we remove the explicit checks for the regular states in the other methods?

Reply:
Which method exactly? The `IsInRegularUnassignState` has to explicitly check that it is not a opt-in feature state, for the others I don't see an explicit check. If it is the `GetNotificationState` I adapted it with the previous commit.

Reply:
You are right, maybe I overlooked something.